### PR TITLE
Rewrite of query fusion and optimization

### PR DESCRIPTION
**Pull request by szeiger**
This is one of the main features for 3.1. This first version contains a new join rewriting that ensures the correct translation of monadic joins.

The later stages still use the old implementation (much of which is not needed anymore) and the produced code is probably worse than before (because all joins are explicit now; the old optimizer only fuses implicit joins).

### CODE REVIEW

File: slick-testkit/src/main/scala/com/typesafe/slick/testkit/tests/NewQuerySemanticsTest.scala [hunk 1]
```diff
@@ -268,7 +268,7 @@ class NewQuerySemanticsTest extends AsyncTest[RelationalTestDB] {
       c <- coffees.filter(_.price < 800).map((_, 1))
     } yield (c._1.name, c._1.supID, c._2)
 
-    val a5 = seq(
+    def a5 = seq(
```
Comment (cvogt) [inline slick-testkit/src/main/scala/com/typesafe/slick/testkit/tests/NewQuerySemanticsTest.scala]: why do all these vals become defs now?

**Comment by cvogt** [comment]
Would be nice if we could somehow test the structure of the generated SQL to make sure changes in the optimizer don't break fusion, etc.

**Comment by nafg** [comment]
In general it would be nice if there were ways for application unit tests
to make assertions about various properties of compiled query ASTs. (Not
sure if that has anything to do with what you said.)

File: slick-testkit/src/main/scala/com/typesafe/slick/testkit/tests/NewQuerySemanticsTest.scala (hunk 1 above)
Comment (szeiger) [inline slick-testkit/src/main/scala/com/typesafe/slick/testkit/tests/NewQuerySemanticsTest.scala]: To get useful logs when an execution fails. Otherwise the several other queries have already been compiled in between.

**Comment by szeiger** [comment]
Checking certain properties of the AST is not hard. The difficult part is knowing what to look for.

**Comment by cvogt** [comment]
Sub query nesting depth in the final SQL could be one metric I think. Number of synthetic column name aliases could be another. Maybe not for this PR, but for 3.1. Even cooler would be if you could write unit tests that send a Slick query to your DBMS, return the describe, parse it and allow checks on it.

**Comment by szeiger** [comment]
Here's an update. All TestKit tests pass on H2. You can see an example of validating AST properties in a unit test in https://github.com/slick/slick/blob/a9ec0598e53032ccaaaf96408baadfde97364ad2/slick-testkit/src/main/scala/com/typesafe/slick/testkit/tests/NewQuerySemanticsTest.scala#L552-562. Aggregation handling is still a hack. I have a more principled approach in mind that I want to try next week.

**Comment by cvogt** [comment]
@szeiger awesome, I like that!

**Comment by szeiger** [comment]
Updated with new aggregation code. I didn't have enough time this week, so it's only at the level of the old code (passes H2 tests). Still needs more work to be able to lift aggregates into joins and unify multiple aggregations of the same source.

**Comment by szeiger** [comment]
Rebased on top of logging improvements. This new version also lifts aggregates into joins. The next step will be proper integration of ROWNUM, zip joins and pagination.

**Comment by szeiger** [comment]
Major milestone reached: After several months of rewriting from scratch, the new back-end passes all tests. Go forth and test it :)

**Comment by cvogt** [comment]
super awesome. way to go @szeiger  :D

**Comment by szeiger** [comment]
Current state: All tests pass, AST code cleaned up considerably, typing edge cases removed. Now working on unified approach to driver-specific pagination / rownum implementations based on first translating everything to zipWithIndex-style operations. Still to do: Extra optimization phase for reordering view operations, possibly a better way of ensuring ordering is never lost. Fixes #489.

**Comment by szeiger** [comment]
Current state: zip and zipWithIndex unified with row number-based pagination, all done in a clean way with AST transformations before merging. The `Subquery` nodes that prevent merging above and below `RowNumber` mappings are used conservatively, thus creating unnecessary subqueries (but the old way of using kludges in the code generator also did that). With the new extra type discipline and testing of `removeTakeDrop`, the code should be very robust. Next step: Less conservative Subquery conditions to avoid subqueries.

**Comment by szeiger** [comment]
Since this version should be free of regressions, I intend to merge the PR after the weekend. 3.1.0-M1 to be released next week (before I go on vacation).
